velox/functions/prestosql/aggregates/EntropyAggregates.cpp

```diff
@@ -0,0 +1,401 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/Aggregate.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/functions/prestosql/aggregates/AggregateNames.h"
+#include "velox/vector/ComplexVector.h"
+#include "velox/vector/DecodedVector.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::aggregate::prestosql {
+
+namespace {
+
+// Indices into Row Vector, in which we store necessary accumulator data.
+constexpr int32_t kSumCIdx{0};
+constexpr int32_t kSumCLogCIdx{1};
+
+// Structure storing necessary data to calculate entropy aggregations.
+struct EntropyAccumulator {
+  EntropyAccumulator() = default;
+
+  EntropyAccumulator(double sumC, double sumCLogC)
+      : sumC_(sumC), sumCLogC_(sumCLogC) {}
+
+  double sumC() const {
+    return sumC_;
+  }
+
+  double sumCLogC() const {
+    return sumCLogC_;
+  }
+
+  void update(int64_t count) {
+    VELOX_CHECK_GT(count, 0, "Entropy count value must be non-negative");
+    if (count == 0) {
+      return;
+    }
+    sumC_ += (double)count;
+    sumCLogC_ += (double)count * std::log(count);
+  }
+
+  void merge(double sumCOther, double sumCLogCOther) {
+    sumC_ += sumCOther;
+    sumCLogC_ += sumCLogCOther;
+  }
+
+  double result() {
+    double entropy = 0.0;
+    if (sumC() > 0.0) {
+      entropy = std::max(
+          (std::log(sumC()) - sumCLogC() / sumC()) / std::log(2.0), 0.0);
+    }
+    return entropy;
+  }
+
+ private:
+  double sumC_{0};
+  double sumCLogC_{0};
+};
+
+// T is the input type for partial aggregation. Not used for final aggregation.
+template <typename T>
+class EntropyAggregate : public exec::Aggregate {
+ public:
+  explicit EntropyAggregate(TypePtr resultType) : exec::Aggregate(resultType) {}
+
+  int32_t accumulatorAlignmentSize() const override {
+    return alignof(EntropyAccumulator);
+  }
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(EntropyAccumulator);
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    setAllNulls(groups, indices);
+    for (auto i : indices) {
+      new (groups[i] + offset_) EntropyAccumulator();
+    }
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    decodedRaw_.decode(*args[0], rows);
+    if (decodedRaw_.isConstantMapping()) {
+      if (!decodedRaw_.isNullAt(0)) {
+        auto value = decodedRaw_.valueAt<T>(0);
+        rows.applyToSelected(
+            [&](vector_size_t i) { updateNonNullValue(groups[i], value); });
+      }
+    } else if (decodedRaw_.mayHaveNulls()) {
```

Comment:
The path the nullable inputs are not covered in test, can you add some tests?  Also the path on line 116

Reply:
It seems that the condition of `!exec::Aggregate::numNulls_` will not be true, because it will call setAllNulls in `initializeNewGroups` methods, so I remove this branch. I don't know if my understanding is correct.
I always refer to Average agg to implement other agg functions, according to my understanding, this branch in Average agg will not be entered too.

Reply:
`!numNulls_` is possible when first batch covers all the groups, decreasing `numNulls_` to 0, and we add second batch, in this case `numNulls_` is 0.

Although in your case, `EntropyAccumulator::update` already has 2 branches, so the optimization probably only matters when `decodedRaw_.isIdentityMapping()` is true, and reducing one branching when `!numNulls_` does not matter too much.